### ISSUE TITLE
Charge scaling in electrostatic interaction depending on the atom indices

### ISSUE THREAD

**Issue by praban07** [openmm/openmm#4151]
Hi,

I learned how to use scaled charges for a subset of atoms from my previous question. I have another issue that requires charge scaling depending on the two indices involved in Coulombic interaction. Here are the details:

Total number of atoms=N;  There are two classes of atoms A and B within N having numbers NA and NB, respectively.
Hence, NA+NB=A

Now, electrostatic interaction consists of two indices (i and j).

case 1:  for i=1, N and j belong to NA, there will be a scaling of Coulombic interaction.
case 2:  for i=1, N and j belong to NB, there will not be any scaling.

However, this cannot be achieved by simply scaling the charges of NA atoms as in case 2 there will be scaling (as NA is a subset of N).

This can be done, in principle, by an if statement which checks whether 'j' index is part of NA or NB.

How to do it in an openMM script? 

I shall greatly appreciate any response. I shall appreciate if you can share a script that does something like that.

**Comment by peastman** [comment]
This is more complicated.  What nonbonded method do you use?

With PME this is very challenging.  PME works by spreading charge onto a grid, then computing the interaction of that charge distribution with itself.  The grid doesn't maintain any record of what atom each bit of charge came from, so you can't modify the interaction based on the pair of atoms.

If you don't need PME it's more straightforward.  The easiest way is to use a CustomNonbondedForce, and define a per-particle parameter that tells whether scaling should be applied.

The description you gave is a little confusing.  When atoms i and j interact, you indicated that scaling should be based on the identity of atom j, but not the identity of atom i.  But it's arbitrary which atom you call i and which you call j.  The interaction needs to treat the two atoms symmetrically.

**Comment by praban07** [comment]
Thanks, Peter. Let me think and rephrase the problem. I shall be back in 1-2 days.

**Comment by praban07** [comment]
Let me rephrase the problem. I would like to modify only the electrostatic part with PME in an explicit solvent MD simulation of biomolecules.

I divide the total number of atoms, N into two groups, P & Q. Now, in the pairwise electrostatic interaction, if one of the indices is from the P group, the interaction is scaled by a constant factor C. However, this cannot be achieved just by scaling the charges of the P group atoms. Because, in that case, if we scale the charges of the P group then

PQ interaction = C*Charge_P * Charge_Q/distance   --- This is correct

PP interaction = C*charge_P * C_charge/distance --- here C comes twice, hence not correct.

One way could be that we scale the P charges and then add a custom non-bonded force only for the P group having the value 1/C.

There could be better ways.  Help will be appreciated.

**Comment by peastman** [comment]
That's doable, but it's going to be a bit expensive.  Take a look at [this cookbook entry](https://openmm.github.io/openmm-cookbook/latest/notebooks/cookbook/Computing%20Interaction%20Energies.html), which discusses how to compute interaction energies.  It performs multiple energy evaluations to separate out the interaction between two groups of atoms from the internal energy of each group.  Once you've done that, you can easily recombine them to get the energy you want.  Something like `scaled_energy = C*total_energy + (1-C)*qq_energy`.

But I assume you want to do this for forces, not just the energy?  You want your simulation to be based on the scaled energy.  You can do that with a CustomCVForce.  Add two NonbondedForces to it, one for the total Coulomb energy and one for the QQ Coulomb energy.  Then you can have it combine them as above.  Since you don't want LJ interactions to be scaled, I would put them in their own NonbondedForce which is *not* inside the CustomCVForce.

**Comment by peastman** [comment]
Note that you don't need to do the manipulations with parameter offsets found in the cookbook entry.  That's so it can evaluate all the different subsets with a single NonbondedForce.  Since you'll have two different NonbondedForces, each one can just have the parameters set correctly.

**Comment by praban07** [comment]
I found out that I just need to do the following:

(a) Use scaled charges for the P group. However, this is not enough as this will overscreen the P-P interaction. Hence, the following step is needed.

(b) Adding an energy term ('force' in openmm terminology) = q_scaled*q_scaled/r  for the P-P interaction.

For (a) Peastman already send a code, for (b) I need to use CustomCVForce. I haven't done (b) yet. I just implemented CustomNonbondedForce (to check if my syntax is working) and plan to change it to CustomCVForce. My code is given at the end. It is running but I will still appreciate it if someone can point out anything I missed.

However, on going from CustomNonbondedForce to CustomCVForce, I see two issues. 

(1) For the P-P interactions, I shall be having NP(NP-1)/2 pairs (NP=number of atoms in the P group). Do I need to define NP(NP-1)/2 CVs? My expression is simple for each pair - q_scale(1)*q_scale(2)/r12  (1, 2 - atom indices, r12 is the distance between 1 and 2).

(2) It seems that 'addPerParicleParameter' (which I used in CustomNonbondedForce to define charges) is not present in CustomCVForce. Then how to add charges? If anyone has an example script that will be very helpful.


Current Code

from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME,
nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond,
0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]
scale_indices=[1,2,3,4]
scale_factors=[2,2,2,2]
for index, scale in zip(scale_indices, scale_factors):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    print('charge',charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge*scale, sigma, epsilon)

scale_map =  dict(zip(scale_indices, scale_factors))
for i in range(nonbonded.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    if p1 in scale_map:
        chargeProd *= scale_map[p1]
    if p2 in scale_map:
        chargeProd *= scale_map[p2]
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

# adding customnonbonded term
el_term=CustomNonbondedForce("ONE_4PI_EPS0*chargesq*(r^-1);chargesq=charge1*charge2;ONE_4PI_EPS0=138.935456")
el_term.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
el_term.setCutoffDistance(nonbonded.getCutoffDistance())
el_term.addPerParticleParameter("charge")
for index in range(nonbonded.getNumParticles()):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    if index in scale_indices:
       charge=charge*scale
       print('scale',scale)
    el_term.addParticle([charge])
system.addForce(el_term)
# adding exceptions

for i in range(nonbonded.getNumExceptions()):
    (p1, p2, q,sigma,epsilon ) = nonbonded.getExceptionParameters(i)
    el_term.addExclusion(p1, p2)
    nonbonded.setExceptionParameters(i, p1, p2, q,sigma,epsilon)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True,
potentialEnergy=True, temperature=True))
simulation.step(10000)

**Comment by peastman** [comment]
You only need two CVs.  Each is a NonbondedForce.  One computes the Coulomb energy for the whole system, and the other computes just the energy of the Q group (set charge to 0 for anything in the P group).  The CustomCVForce is just to combine those two in a way that scales the energy of anything involving a P atom, but not things that don't.

You don't need to scale the charges.  That just complicates things.  Instead, the CustomCVForce will scale the energy.

**Comment by praban07** [comment]
Apologies for my limited knowledge of openmm and CustomCVForce. Isn't it true that the CV should be a set of variables? If so, can a nonbonded force be a CV? The collection of all pairwise distances can be taken as CV and using that one can calculate the Coulomb energy. Did you mean that? 

Actually, I was looking for some examples of using CV - got only one (https://openmm.github.io/openmm-cookbook/latest/notebooks/tutorials/umbrella_sampling.html?highlight=customcvforce). If there is an example of using CV, that is a set of multiple variables, that will be very helpful to me. Thanks.

**Comment by peastman** [comment]
Any Force can act as a CV.  CustomCVForce takes the energies they return, evaluates a user defined function of them, and returns that as the final energy.  Take this trivial example.

```python
cv = CustomCVForce('x')
cv.addCollectiveVariable('x', force1)
```

This CustomCVForce basically does nothing.  It evaluates the energy of `force1` and returns it.  The result is exactly the same as if you just added `force1` directly to your System.

Here is an example with multiple CVs.


```python
cv = CustomCVForce('x+y')
cv.addCollectiveVariable('x', force1)
cv.addCollectiveVariable('y', force2)
```

This evaluates the energies of two forces and returns the sum.  Which again is exactly the same as if you added them directly to the System.  Things get more interesting when you use more complicated expressions.


```python
cv = CustomCVForce('C*x + (1-C)*y')
cv.addCollectiveVariable('x', force1)
cv.addCollectiveVariable('y', force2)
cv.addGlobalParameter('C', 1.5)
```

Suppose that `force1` is a NonbondedForce that evaluates all interactions in the System, and `force2` is a second NonbondedForce that evaluates only interactions within the Q group.  It begins by scaling the total interaction energy by `C`.  But then it subtracts off the scaled energy `C*y` of the Q group, and adds back in the unscaled energy `1*y`.  The result is that interactions within the Q group are unscaled, and all other interactions are scaled by `C`.

**Comment by praban07** [comment]
I have given my code below (added only one force). But still has a few doubts. Are the following statements correct?

(1) It seems we have to add the electrostatic term explicitly as 'NonbondedForce' has both LJ + Coulomb. There is no way to extract Coulomb interaction from the NonbondedForce and then give that as a force in CustomCVForce.

(2) I have made the Coulombic force using CustomNonbondedForce and then used that force for CustomCVForce. Is it the correct procedure?

(3) I am not absolutely sure if I have defined everything correctly so that my new force ('el_term') is going to give the exact value as the Coulomb part of the forcefield.  Is there a way to write these two separately?

(4) Finally is the code given below correct? It runs.

------------------------------------
from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

# defining my force in terms of CustomNonbondedForce
el_term=CustomNonbondedForce("ONE_4PI_EPS0*chargesq*(r^-1);chargesq=charge1*charge2;ONE_4PI_EPS0=138.935456")
el_term.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
el_term.setCutoffDistance(nonbonded.getCutoffDistance())
el_term.addPerParticleParameter("charge")
for index in range(nonbonded.getNumParticles()):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    el_term.addParticle([charge])

# Adding exceptions
for i in range(nonbonded.getNumExceptions()):
    (p1, p2, q,sigma,epsilon ) = nonbonded.getExceptionParameters(i)
    el_term.addExclusion(p1, p2)
    nonbonded.setExceptionParameters(i, p1, p2, q,sigma,epsilon)

# now definding CustomCVForce with the force already defined

cv=CustomCVForce('x')
cv.addCollectiveVariable("x",el_term)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)

**Comment by peastman** [comment]
You don't need to use a CustomNonbondedForce.  You can use a NonbondedForce and just set epsilon to 0 for every particle.  That way you can still use PME.

The CustomNonbondedForce you've written applies a sharp cutoff.  The energy drops abruptly to 0 at 1 nm.  That isn't a very accurate way of computing Coulomb interactions.  NonbondedForce uses more accurate methods.

You build a CustomNonbondedForce that is meant to reproduce the Coulomb interaction from the NonbondedForce.  But you don't remove or modify the NonbondedForce.  It's still in the System and still computes the interactions.  That means the Coulomb interactions will get included twice, once by the NonbondedForce and once by the CustomCVForce.

**Comment by praban07** [comment]
I have tried this now. I had the intention to subtract 0.5*Coulomb force (that line is currently commented out as I am getting other errors).

I am getting the following error message. 

Traceback (most recent call last):
  File "/home/pradipta/OPENMM/RUN/CV_elec_sim_AMBER.py", line 31, in <module>
    cv.addCollectiveVariable('x',nonbonded)
  File "/home/pradipta/anaconda3/envs/openmm_cuda/lib/python3.11/site-packages/openmm/openmm.py",
line 14362, in addCollectiveVariable
    raise Exception(s)
Exception: the CustomCVForce object does not own its corresponding OpenMM object

The code is given below

---------------

from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

#nonbonded=-0.5*nonbonded


for index in range(nonbonded.getNumParticles()):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    epsilon=0
    nonbonded.addParticle(charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge, sigma, epsilon)

# Adding exceptions
for i in range(nonbonded.getNumExceptions()):
    (p1, p2, q,sigma,epsilon ) = nonbonded.getExceptionParameters(i)
#    el_term.addExclusion(p1, p2)
    nonbonded.setExceptionParameters(i, p1, p2, q,sigma,epsilon)

# now defining CustomCVForce with the force already defined

cv=CustomCVForce('x')
cv.addCollectiveVariable('x',nonbonded)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)
------------------------------------

**Comment by praban07** [comment]
Now it is running. All I needed to do is 

(1) scale the charges of atoms of a group (say P). I have taken only 2 atoms to be part of P as a prototype.
(2) Step (1) introduces extra scaling of atoms, when both elements of a pair of atoms belongs to P. So, added
    a customCVForce to nullify the extra scaling.

I would appreciate if any issue in the code is pointed out.

The codes starts below:

from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)
nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

scale_indices=[1,2]
scale_factors=[0.5,0.5]
for index, scale in zip(scale_indices, scale_factors):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    print('charge',index,charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge*scale, sigma, epsilon)

scale_map =  dict(zip(scale_indices, scale_factors))
for i in range(nonbonded.getNumExceptions()):
p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    if p1 in scale_map:
        chargeProd *= scale_map[p1]
    if p2 in scale_map:
        chargeProd *= scale_map[p2]
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)



el_term=copy.copy(nonbonded)

for index in range(el_term.getNumParticles()):
    charge, sigma, epsilon = el_term.getParticleParameters(index)
    if index in scale_indices:
       print('charge-after',index,charge,sigma,epsilon)
       epsilon=0
   else:
       charge=0
       epsilon=0

    el_term.setParticleParameters(index, charge, sigma, epsilon)


cv=CustomCVForce('x')
cv.addCollectiveVariable('x',el_term)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)
--The code ends----

**Comment by peastman** [comment]
Can you edit your message and put lines with \`\`\` before and after the code?  It's getting rendered as markdown.  See https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks.

**Comment by praban07** [comment]
I did. Please see my previous message.

**Comment by peastman** [comment]
```python
cv=CustomCVForce('x')
cv.addCollectiveVariable('x',el_term)
```

Your CustomCVForce isn't doing anything useful.  All it does is return the energy of the force it contains directly.  You can eliminate it and just add the NonbondedForce directly to the System.

You're using a different approach from what I suggested: scaling the charges rather than scaling the total energy.  That can work, but with a limitation.  It can only be used to decrease the strength of the interactions, not to increase it.  If you scale the charges of the P atoms by $x$, you scale P-P interactions by $x^2$ and all other interactions by $x$.  Then you add a second NonbondedForce that scales the charges of P atoms by $y$, thus scaling P-P interactions by $y^2$, and sets all other charges to 0.  You want the sum of the two forces to give P-P interactions scaled by $x$.  That is, you want $x^2+y^2=x$.  Solving for $y$ gives $y=\sqrt{x-x^2}$.  That has solutions for $x<1$, but not for $x>1$.  As long as you're ok with that limitation, this approach can work.

**Comment by praban07** [comment]
Here is the code with energy scaling. Certainly it is more flexible. Only issue is adding CV force seems to make the simulation slower than adding a nonbonded force.
``````````````````````````````````````````````````````````````````````````````````````
from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

el_term_1=copy.copy(nonbonded)
el_term_2=copy.copy(nonbonded)

scale_indices=[1,2]



for index in range(el_term_2.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    if index in scale_indices:
       charge=0
       epsilon=0
    else:
       epsilon=0

    el_term_2.setParticleParameters(index, charge, sigma, epsilon)


cv=CustomCVForce('C*x+(1-C)*y')
cv.addCollectiveVariable('x',el_term_1)
cv.addCollectiveVariable('y',el_term_2)
cv.addGlobalParameter('C',0.5)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)
````````````````````````````````````````````````````````````````````````````````````
"energy_scale.py" 45L, 1405C                                                                                                  34,1          Top
````````````````````````````````````````````````````````````````````````````````````````````````````````

**Comment by peastman** [comment]
You add a CustomCVForce that computes the electrostatic energy, but the original NonbondedForce is still present and computing it as well.  In that original NonbondedForce, you should set the charge of every particle to 0 so it will only compute the LJ part.

You also need to adjust the exceptions in the NonbondedForces.  Presumably you want them to be handled in the same way as other interactions: scale the electrostatic part by C and leave the LJ part unchanged?  The easiest way of doing that is to keep them in the original NonbondedForce, multiplying `chargeProd` by C if either particle is in set P.  In `el_term_1` and `el_term_2`, set `chargeProd` and `epsilon` to 0 for every exception.

**Comment by praban07** [comment]
Corrected the code. Hopefully it is fully correct now.
````````````````````````````````````````````````````````````````````````````````````````````````````````````

from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

scale_indices=[1,2]
scale_factors=[0.5,0.5]
zero=0

# the following block ensures that the original nonbonded force calculates only
# LJ, not the electrostatics. All charges are set to zero.
    
for index, scale in zip(scale_indices, scale_factors):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    print('charge',charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge*zero, sigma, epsilon)
    

scale_map =  dict(zip(scale_indices, scale_factors))

# putting exceptions in the nonbonded force with scaled chargeProd

for i in range(nonbonded.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    if p1 in scale_map:
        chargeProd *= scale_map[p1]
    if p2 in scale_map:
        chargeProd *= scale_map[p2]
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

el_term_1=copy.copy(nonbonded)
el_term_2=copy.copy(nonbonded)




for index in range(el_term_2.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    if index in scale_indices:
       charge=0
       epsilon=0
    else:
       epsilon=0

    el_term_2.setParticleParameters(index, charge, sigma, epsilon)

# taking care of exceptions in el_term_1 and el_term_2

for i in range(el_term_1.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_1.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_1.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

for i in range(el_term_2.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_2.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_2.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)
a, epsilon = nonbonded.getParticleParameters(index)
# now form the CV
cv=CustomCVForce('C*x+(1-C)*y')
cv.addCollectiveVariable('x',el_term_1)
cv.addCollectiveVariable('y',el_term_2)
cv.addGlobalParameter('C',0.5)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)
``````````````````````````````````````````````````````````````````````````````````````````````````````````````````

**Comment by praban07** [comment]
Corrected the code. Hope this is correct.
```````````````````````````````````````````````````````````````````````````````
from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]

scale_indices=[1,2]
scale_factors=[0.5,0.5]
zero=0

el_term_1=copy.copy(nonbonded)
el_term_2=copy.copy(nonbonded)

# the following block ensures that the original nonbonded force calculates only
# LJ, not the electrostatics. All charges are set to zero.

for index, scale in zip(scale_indices, scale_factors):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    print('charge',charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge*zero, sigma, epsilon)

scale_map =  dict(zip(scale_indices, scale_factors))

# putting exceptions in the nonbonded force with scaled chargeProd

for i in range(nonbonded.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    if p1 in scale_map:
        chargeProd *= scale_map[p1]
    if p2 in scale_map:
        chargeProd *= scale_map[p2]
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)



for index in range(el_term_1.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    epsilon=0

    el_term_1.setParticleParameters(index, charge, sigma, epsilon)


for index in range(el_term_2.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    if index in scale_indices:
       charge=0
       epsilon=0
    else:
       epsilon=0

    el_term_2.setParticleParameters(index, charge, sigma, epsilon)

# taking care of exceptions in el_term_1 and el_term_2

for i in range(el_term_1.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_1.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_1.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

for i in range(el_term_2.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_2.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_2.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

# now form the CV
cv=CustomCVForce('C*x+(1-C)*y')
cv.addCollectiveVariable('x',el_term_1)
cv.addCollectiveVariable('y',el_term_2)
cv.addGlobalParameter('C',0.5)

system.addForce(cv)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter('output.pdb', 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))
simulation.step(10000)
```````````````````````````````````````````````````````````````````````````````````````````````````````````

**Comment by peastman** [comment]
I'm not sure this code does what you want:

```python
    if p1 in scale_map:
        chargeProd *= scale_map[p1]
    if p2 in scale_map:
        chargeProd *= scale_map[p2]
```

If both particles are in the scaled group, you'll scale the interaction by $C^2$ instead of $C$.

Once you think you have everything correct, you can run a few tests to check for problems.

- Set the scale factor to 1.  The energy should come out identical to the original system.
- Make the scaled group include every particle in the entire system.  In that case, it should just scale the total electrostatic energy by $C$.
- Set the scale factor to 0.  Compare to a system in which their charges (including in exceptions) are all 0.  It should match.

If there's any problem in your implementation, there's a good chance that one of those tests will catch it.

**Comment by praban07** [comment]
Now it works. 

It passed all three tests you mentioned. The code is given at the end. 

Outputs for a standard run and scaled one (with scale factor=1) are attached (I did 10 steps of minimization and MD). The total potential energy is the same up to 2nd digit after decimal. I could compare the energies before the start of minimization. Once, the minimization starts they differ; however, even for a standard case, two different runs do not give identical results in minimization (random number seed? Not sure where a random number is used in minimization).

[energy_output_standard.txt](https://github.com/openmm/
[energy_output_scaled_with_1.txt](https://github.com/openmm/openmm/files/12368426/energy_output_scaled_with_1.txt)
openmm/files/12368424/energy_output_standard.txt)

`````````````````````````````````````````````````````
from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('input.prmtop')
inpcrd = AmberInpcrdFile('input.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinMiddleIntegrator(300*kelvin, 1/picosecond, 0.004*picoseconds)

# extracting nonbonded force
nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]
# atoms for which charges to be scaled.
scale_indices=[1,2]
# we just need the first element of the scale factor for scaling
scale_factors=[1.0,0.5]
zero=0
# making two copies of the nonbonded force. These will be used later to define CVForce
el_term_1=copy.copy(nonbonded)
el_term_2=copy.copy(nonbonded)

# the following block ensures that the original nonbonded force calculates only
# LJ, not the electrostatics for all particles. All charges are set to zero.

for index in range(nonbonded.getNumParticles()):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
#    print('charge',charge,sigma,epsilon)
    nonbonded.setParticleParameters(index, charge*zero, sigma, epsilon)

#nonbonded.setForceGroup(1)
scale_map =  dict(zip(scale_indices, scale_factors))

# putting exceptions in the nonbonded force with scaled chargeProd

for i in range(nonbonded.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    #print('p1p2-all',p1,p2)
# is either p1 or p2 or both are present, we scale the chargeprod
    if ((p1 and p2)or(p1 or p2)) in scale_map:
    #    print('p1p2',p1,p2)
        chargeProd *= scale_factors[0]
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)


# el_term_1 calculates only electrostatics for all particles
for index in range(el_term_1.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    epsilon=0

    el_term_1.setParticleParameters(index, charge, sigma, epsilon)

# el_term_2 calculates electrostatics for only the atoms, which are not scaled
for index in range(el_term_2.getNumParticles()):
    charge, sigma, epsilon = el_term_2.getParticleParameters(index)
    if index in scale_indices:
       charge=0
       epsilon=0
    else:
       epsilon=0

    el_term_2.setParticleParameters(index, charge, sigma, epsilon)

# taking care of exceptions in el_term_1 and el_term_2

for i in range(el_term_1.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_1.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_1.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

for i in range(el_term_2.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = el_term_2.getExceptionParameters(i)
    chargeProd = 0
    epsilon=0
    el_term_2.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)

# now form the CV
#cv=CustomCVForce('C*x+(1-C)*y')
cv=CustomCVForce('C*x')
cv.addCollectiveVariable('x',el_term_1)
cv.addCollectiveVariable('y',el_term_2)
cv.addGlobalParameter('C',1.0)


# adding CV force
system.addForce(cv)

# calling setForceGroup

for i, f in enumerate(system.getForces()): 
            f.setForceGroup(i)

simulation = Simulation(prmtop.topology, system, integrator)
simulation.context.setPositions(inpcrd.positions)
if inpcrd.boxVectors is not None:
    simulation.context.setPeriodicBoxVectors(*inpcrd.boxVectors)
sum=0.0
# get the energy component before starting minimization
for i, f in enumerate(system.getForces()):
        state = simulation.context.getState(getEnergy=True, groups={i})
        print(f.getName(), state.getPotentialEnergy())
#        energies[f] = state.getPotentialEnergy()  
print("Total potential energy ",simulation.context.getState(getEnergy=True).getPotentialEnergy())

# now for debugiing running Minimization for 10 steps and printing energies
for minStep in range(10):
    print("minimization step: ",minStep+1)
    simulation.minimizeEnergy( maxIterations=1)
    for i, f in enumerate(system.getForces()):
        state = simulation.context.getState(getEnergy=True, groups={i})
        print(f.getName(), state.getPotentialEnergy(),end=",")
    print("Total potential energy ",simulation.context.getState(getEnergy=True).getPotentialEnergy())

simulation.reporters.append(PDBReporter('output.pdb', 1000))
###  RUNNING MD FOR 10 steps and writing energies
for mdStep in range(10):
    print("MD step: ",mdStep+1)
    simulation.step(1)
    for i, f in enumerate(system.getForces()):
        state = simulation.context.getState(getEnergy=True, groups={i})
        print(f.getName(), state.getPotentialEnergy(),end=",")
    print("Total potential energy ",simulation.context.getState(getEnergy=True).getPotentialEnergy())

simulation.reporters.append(PDBReporter('output.pdb', 1000))
````````````````````````````````````````````````````````````````````````````````````

**Comment by peastman** [comment]
> if ((p1 and p2)or(p1 or p2)) in scale_map:

I think you meant that to be `p1 in scale_map or p2 in scale_map`?

**Comment by praban07** [comment]
You are right. Corrected it. Previously it was missing some pairs having p1 or p2.

How to reproduce the results of minimization in openMM (say, for a standard AMBER forcefield)?

**Comment by peastman** [comment]
Do you mean how to make minimization deterministic?  I believe it should generally be deterministic with the CPU and OpenCL platforms, and CUDA will be if you specify the property `DeterministicForces`:`true` when creating the Simulation.

```python
simulation = Simulation(prmtop.topology, system, integrator, Platform.getPlatformByName('CUDA'), {`DeterministicForces`:`true`})
```

**Comment by praban07** [comment]
Now, I am trying to scale the LJ parameters. I have tried for the scaling 'sigma'. Following is the relevant part of the code. It runs. Just want to know if it is fine.
``````````````````````````````````````````````````
from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout

prmtop = AmberPrmtopFile('/home/pradipta/OPENMM/CaM/loop1_w.prmtop')
inpcrd = AmberInpcrdFile('/home/pradipta/OPENMM/CaM/loop1_w.inpcrd')
system = prmtop.createSystem(nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)


# extracting nonbonded force
nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]
# atoms for which charges to be scaled.
scale_indices=[14,15,48,49,67,68,91,172,173,182]
# the atom(s) whose LJ parameters to modify are listed below
LJ_index=[172,173,182]
# we just need the first element of the scale factor for scaling. The value of the first element must match 'C' in cvforce
scale_factors=[0.5,0.5,1,1,1]
zero=0
# LJ_scale is defined here
LJ_scale=1.2
#
# making two copies of the nonbonded force. These will be used later to define CVForce
el_term_1=copy.copy(nonbonded)
el_term_2=copy.copy(nonbonded)

# the following block ensures that the original nonbonded force calculates only
# LJ, not the electrostatics for all particles. All charges are set to zero.
# LJ is scaled for LJ_index atoms

LJset = {}
for index in range(nonbonded.getNumParticles()):
    charge, sigma, epsilon = nonbonded.getParticleParameters(index)
    LJset[index]=(sigma,epsilon)
#    print('charge',index,charge,sigma,epsilon)
    if index in LJ_index:
        sigma=sigma* LJ_scale
    nonbonded.setParticleParameters(index, charge*zero, sigma, epsilon)


scale_map =  dict(zip(scale_indices, scale_factors))

for i in range(nonbonded.getNumExceptions()):
    p1, p2, chargeProd, sigma, epsilon = nonbonded.getExceptionParameters(i)
    val_1=LJset[p1][0]
    val_2=LJset[p2][0]
    
# is either p1 or p2 or both are present, we scale the chargeprod
    if p1 in scale_indices or p2 in scale_indices:
#        print('p1p2',p1,p2)
        chargeProd *= scale_factors[0]
    if p1 in LJ_index:
        val_1 *= LJ_scale
    if p2 in LJ_index:
        val_2 *= LJ_scale

    sigma=0.5*(val_1+val_2)
    nonbonded.setExceptionParameters(i, p1, p2, chargeProd, sigma, epsilon)
``````````````````````````````````````````````````````````````````````````````````````````````````````

**Comment by peastman** [comment]
I don't see any obvious problems.   In processing the exceptions, you ignore the existing sigma value and replace it with a new value computed from the per-particle parameters.  Usually that should be ok, because force fields typically modify epsilon but not sigma for the exceptions.  Make sure it's really producing the values you want it to have.  If the existing exceptions have modified sigma values, those modifications will get thrown out.

**Comment by praban07** [comment]
I think it is working. I don't have any previous results to compare, at this moment. However, the trends are correct.

So, if the forcefield (AMBER14SB in my case) modifies only epsilon for exceptions, then I don't need to modify the exception when I change only sigma. Right?

**Comment by peastman** [comment]
You still need to change sigma if you want to scale sigma.  Any interaction listed in an exception will use the parameters in the exception.  If you don't change the sigma in the exception, it will continue to use the old value.

**Comment by praban07** [comment]
Yes, you are correct. I misunderstood.

You can close this thread. I can run what I wanted. openmm has made it quite easy. Your help has been tremendous. Thanks a ton. For other type of questions, I shall open new issues.
